src/coreclr/vm/exceptionhandling.cpp

```diff
@@ -3002,23 +3002,70 @@ extern "C" void QCALLTYPE AppendExceptionStackFrame(QCall::ObjectHandleOnStack e
     END_QCALL;
 }
 
-#if defined(HOST_AMD64) && defined(HOST_WINDOWS)
-size_t GetSSPForFrameOnCurrentStack(TADDR ip)
+void ExecuteFunctionBelowContext(PCODE functionPtr, CONTEXT *pContext, size_t targetSSP, size_t arg1, size_t arg2)
 {
-    size_t *targetSSP = (size_t *)_rdsspq();
-    // The SSP we search is pointing to the return address of the frame represented
-    // by the passed in IP. So we search for the instruction pointer from
-    // the context and return one slot up from there.
-    if (targetSSP != NULL)
+    UINT_PTR targetSp = GetSP(pContext);
+#if defined(HOST_AMD64)
+    ULONG64* returnAddress = (ULONG64*)(targetSp - 8);
+    *returnAddress = pContext->Rip;
+#ifdef HOST_WINDOWS
+    if (targetSSP != 0)
     {
-        while (*targetSSP++ != ip)
-        {
-        }
+        targetSSP -= sizeof(size_t);
+    }
+#endif // HOST_WINDOWS
+    SetSP(pContext, targetSp - 8);
+#elif defined(HOST_X86)
+
+#ifdef HOST_WINDOWS
+    // Disarm the managed code SEH handler installed in CallDescrWorkerInternal
+    if (IsCallDescrWorkerInternalReturnAddress(pContext->Eip))
+    {
+        PEXCEPTION_REGISTRATION_RECORD currentContext = GetCurrentSEHRecord();
+        if (currentContext->Handler == (PEXCEPTION_ROUTINE)ProcessCLRException)
+            currentContext->Handler = (PEXCEPTION_ROUTINE)CallDescrWorkerUnwindFrameChainHandler;
     }
+#endif
 
-    return (size_t)targetSSP;
+    ULONG32* returnAddress = (ULONG32*)(targetSp - 4);
+    *returnAddress = pContext->Eip;
+    SetSP(pContext, targetSp - 4);
+#elif defined(HOST_ARM64)
+    pContext->Lr = GetIP(pContext);
+#elif defined(HOST_ARM)
+    pContext->Lr = GetIP(pContext);
+#elif defined(HOST_RISCV64) || defined(HOST_LOONGARCH64)
+    pContext->Ra = GetIP(pContext);
+#endif
+
+// The SECOND_ARG_REG is defined only for Windows, it is used to handle longjmp propagation over managed frames
+#ifdef HOST_AMD64
+#ifdef UNIX_AMD64_ABI
+#define FIRST_ARG_REG Rdi
```

Comment:
Can this be in cgencpu.h and de-duped with the `GetFirstArgReg` that you have added in this change?